### ISSUE TITLE
Link `focus` and `hover` should follow WAI-ARIA WCAG AA guidelines

### ISSUE THREAD

**Issue by summercms** [openui/open-ui#658]
Not sure if this is something this repo can use, but decided to open an issue.

### Code Example 1

```css
a {
    color: #2AB1E3;
}
a:focus, a:hover {
    color: #1892BF;
}
```

Gives the following result:

![image](https://user-images.githubusercontent.com/57409060/215775259-65136fa2-26dc-402d-a5e8-161621b88bfc.png)

The above makes it hard to tell if the website is confirming you are **hovering** over a link with bad eye sight.

### Code Example 2

```css
a {
    color: #2AB1E3;
}
a:focus, a:hover {
    color: #083644;
}
```

Gives the following result:

![image](https://user-images.githubusercontent.com/57409060/215775695-9a5c1892-fa8e-4c2e-8955-801fbf3038d3.png)

The above makes it obvious the website is confirming you are **hovering** over a link with poor eye sight.

### W3c Spec

Did a search and can find info on `focus` ring. But no mention above `hover`

Link: https://www.w3.org/WAI/standards-guidelines/wcag/new-in-22/#2411-focus-appearance-aa

**Comment by summercms** [comment]
@gregwhitworth 

Here's some ideas to get the ball rolling.

### Goals

1. Browsers add a test to their dev tools section to flag any impaired vision text link hovering confirmations.
2. Tools like Lighthouse could add a rule to check impaired vision text link hovering confirmations.
3. Repos like this one that are creating web component elements could add this. For example a `table` element that has text links in it's **header columns** (that perform a task) should follow WAI-ARIA WCAG AA guidelines.
4. W3c to add **impaired vision text link hovering confirmations** to it's spec.

### Notes

When checking a `text link hover / focus` to see if the contrast difference is enough or not needed, there could be some basic checks:

- Does the text link add or remove any borders or underlines when hovered / focused etc.
- Is the borders or underlines styling added or removed when hovered and very different, for example: 
  - They add a 1px solid underline that's thin and almost the same colour as the background on hover (would fail).
  - They add a 1px dotted underline that's thin but very different colour to the background on hover (would fail).
  - They add a 5px thick underline on the text link hover (would pass as easy to tell the difference).
  etc.

I welcome people's suggestions to expand this idea.

**Comment by gregwhitworth** [comment]
Thank you @summercms for expounding; let me go one by one.

> Browsers add a test to their dev tools section to flag any impaired vision text link hovering confirmations.

Browsers have begun doing this, for example here is Edge (it should be in any Chromium dev tool however):

<img width="792" alt="image" src="https://user-images.githubusercontent.com/865244/227784969-2e86a053-f8f2-4fc0-8604-dae5ea348ef7.png">

> Tools like Lighthouse could add a rule to check impaired vision text link hovering confirmations

This is a good idea but it's also worth understanding how hard this actually is; focusing on the styles in isolation this is somewhat easy; however you then need to take layout and positioning into account as it may give you a false positive. The only way in which to **ensure** contrast in a test is to output the render tree (which is not standardized) or output it to an image and then do image analysis. Even then though; there are moments where text is used for design reasons so you'd then need to map the recognition of the text from the image to the accessibility tree to derive if that text is not only visible but screen-reader accessible. Due to this complexity, this is why when you enable High Contrast on Windows they actually replace all of the background styles of the text blocks to match the user's theme as this is the only way in which they can truly ensure contrast and maintain a performant experience.

> Repos like this one that are creating web component elements could add this. For example a table element that has text links in it's header columns (that perform a task) should follow WAI-ARIA WCAG AA guidelines.

100% agree with this and if there are any areas where you see that the site isn't matching guidelines, please file an issue. Additionally, while we work on controls/components that we introduce we'll work with the ARIA WG and accessibility experts to ensure that they are accessible by default when they ship.

> W3c to add impaired vision text link hovering confirmations to it's spec.

This seems like an issue I'd file with the CSSWG although they'll hit similar issues as I outlined with the testing as the algo's on styles alone can only go so far. 

![image](https://user-images.githubusercontent.com/865244/227785965-69957c65-cd9f-4c59-a9ae-f87528463f66.png)

We actually did a decent amount of user research when introducing the new focus indicators when Edge moved to Chromium and this is where we discussed how we could possibly **ensure** visibility by default, you can read more about it [here](https://blogs.windows.com/msedgedev/2019/10/15/form-controls-microsoft-edge-chromium/) (scroll down to _More accessible controls_).
